### DISCUSSION TITLE
[Proposal] Allow to assign readonly field inside initializer

### DISCUSSION THREAD

**Discussion by ds1709** [dotnet/csharplang#6193] [Language Ideas]
Allow to assign readonly fields inside class initializer.
```c#
public sealed class Foo
{
    private readonly int bar;

    public static Foo Create(int x)
    {
        return new Foo
        {
            bar = x // CS0191: A readonly field cannot be assigned to (except in a constructor or init-only setter of the type in which the field is defined or a variable initializer)
        };
    }
}
```
Why we can't assign its value right from initializer (especially since `init` property can do it)?
There're some workarounds of it, but all of them has disadvantages:
* `Remove readonly keyword and make field not-readonly`. Maybe field realy must be readonly, and this workaround is unacceptable.
* `Use parametrized constructor`. Static factory methods are usualy used to make hard work to prepare data for class instantiation. As a result of this work, a large number of parameters may appear to initialize large number of fields. So, in this case you need to make constructor with large number of parameters.
* `Make init property over readonly field to initialize it`. If you need this property only to initialize field, then it will be property without get, or whith unused get. In both cases, it's bad type design practice.

**Comment by HaloFour** [comment]
Being able to write to accessible `readonly` fields from an object initializer would be _massively_ dangerous.  The compiler cannot know why the field is `readonly` or the ramifications of mutating it post-construction.  If the developer of the class intended the field to remain mutable during initialization then would expose it as an `init` property.  That's why `init` properties exist, and a write-only properties in general are not bad design.

**Reply by ds1709** [reply]
Do you mean adding `init` property to initialize readonly field is some kind of *explicit* code to make readonly field initializable from class initializer? Correct me if I wrong.
Ok than, can we have some kind of `init` fields, that will work like readonly field, but can be initialized from initializer (same like `init` property do)?
Like this
```c#
public sealed class Foo
{
    private init int bar;
    // or
    private readonly init int bar;

    public static Foo Create(int x)
    {
        return new Foo { bar = x; };
    }
}
```

**Reply by HaloFour** [reply]
Yes, I would be concerned about the intent of the author of that class.  There are tons of classes with `readonly` fields where allowing a consumer to mutate them even only during initialization could have deleterious side effects.

I do think `init` fields would cover that concern, so only specially marked fields can be mutated during initialization.  Question would be if the language team finds them to be worthwhile over a `init` property in cases like this.  A read/init auto-property is almost as concise and doesn't add any overhead thanks to inlining in the runtime.

**Comment by ds1709** [comment]
About set-only properties, there's a guideline on MSDN:
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/property
Also, there're code analysis rules, that warn developer about set-only properties.
So, IMHO, set-only properties is bad type design practice. Using init-only properties to initialize readonly field seems more like *code smell*, than *workaround*.

**Reply by HaloFour** [reply]
Those are only guidelines and apply generally to the public surface of a type.

Those same guidelines also suggest not exposing fields:

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/field

**Reply by ds1709** [reply]
In my scenario field is not exposed, it's private field of class. Also, discussing field design guidelines is offtopic.

**Reply by HaloFour** [reply]
Discussing field design guidelines is as relevant as discussing property design guidelines.

If these fields are private there's nothing stopping you from also adding a new constructor which allows you to set that field.

**Reply by ds1709** [reply]
It's OK if you have a small number of contructor parameters. But if numer of parameters is large (e.g. ten or more), then it's (again) bad type design practice.
PS: property design guideline is relevant coz using init-only property is one of posible workarounds.

**Reply by HaloFour** [reply]
> PS: property design guideline is relevant coz using init-only property is one of posible workarounds.

Those guidelines only apply to accessible members, not private ones.

**Reply by HaloFour** [reply]
Unless you're taking `ref` to any of those fields directly your use case seems perfectively satisfied by using auto-properties:

```cs
public sealed class Foo
{
    private int Bar { init; get; }

    public static Foo Create(int x)
    {
        return new Foo
        {
            Bar = x
        };
    }
}
```

**Reply by ds1709** [reply]
>Those guidelines only apply to accessible members, not private ones.

`init` properties are accesible, even if it must be not. So it's one more disadvantage of using `init` property.

**Reply by ds1709** [reply]
>Unless you're taking ref to any of those fields directly ...

Never tried it, but... can I use `readonly` field as a `ref` parameter?

**Reply by HaloFour** [reply]
> `init` properties are accesible, even if it must be not. So it's one more disadvantage of using `init` property.

Properties can be just as `private` as fields.  Private members don't need to fit the design guidelines, they are implementation details.

**Reply by FaustVX** [reply]
@ds1709 
> can I use readonly field as a ref parameter?

No, `ref` allow the caller to modify the value, which is the opposite of `readonly`
